bundles/org.openhab.binding.fenecon/src/main/java/org/openhab/binding/fenecon/internal/FeneconHandler.java

```diff
@@ -0,0 +1,231 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.library.types.DateTimeType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link FeneconHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Philipp Schneider - Initial contribution
+ */
+@NonNullByDefault
+public class FeneconHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(FeneconHandler.class);
+
+    private FeneconConfiguration config = new FeneconConfiguration();
+    private @Nullable ScheduledFuture<?> pollingJob;
+
+    private final List<String> channels = List.of("GridMode", "State", "EssSoc", "ConsumptionActivePower",
+            "ProductionActivePower", "GridActivePower", "EssDischargePower", "GridBuyActiveEnergy",
+            "GridSellActiveEnergy");
+
+    private final HttpClient httpClient;
+    private @Nullable Builder baseHttpRequest;
+
+    public FeneconHandler(Thing thing) {
+        super(thing);
+        httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(FeneconConfiguration.class);
+
+        logger.info("FENECON: initialize REST-API connection to {} with polling interval: {} sec", getBaseUrl(config),
+                config.refreshInterval);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        baseHttpRequest = createBaseHttpRequest(config);
+        pollingJob = scheduler.scheduleWithFixedDelay(this::pollingCode, 5, config.refreshInterval, TimeUnit.SECONDS);
+    }
+
+    private static final String getBasicAuthHeader(String username, String password) {
+        String valueToEncode = username + ":" + password;
+        return "Basic " + Base64.getEncoder().encodeToString(valueToEncode.getBytes());
+    }
+
+    private String getBaseUrl(FeneconConfiguration config) {
+        return "http://" + config.hostname + ":" + config.port + "/";
+    }
+
+    private Builder createBaseHttpRequest(FeneconConfiguration config) {
+        String basicAuth = getBasicAuthHeader("x", config.password);
+        return HttpRequest.newBuilder().timeout(Duration.ofSeconds(5)).header("Authorization", basicAuth)
+                .header("Content-Type", "application/json").GET();
+    }
```

Comment:
For good seperation of concerns it would be better to extract these and other http/api handling out of this `FeneconHandler` class and move it into it's own `ApiController` or whatever you name it. 

A common approach is to call public methods on that `ApiController` and whenever it fails, throw custom exceptions within that `ApiController`  class. Those can be catched by the calling `FeneconHandler` class that can set the `Thing` state accordingly. 

An example is the PlugwiseHA binding, maybe that can be used as some kind of inspiration.

Reply:
@lsiepel So now everything should be done review-wise 😊